### PR TITLE
vcm-436 adding Card Metadata API

### PR DESCRIPTION
**Pull request by jaak-niit-cko**
# Description of changes in PR

[//]: # 'Please add your description here'
Adding Card Metadata API
Product description - (https://checkout.atlassian.net/wiki/spaces/VLT/pages/5338202294/Public+Card+Metadata+API+-+Stage+1)
Technical description - (https://checkout.atlassian.net/wiki/spaces/VLT/pages/5428445493/Public+Card+Metadata+Public+API+-+Tech+Proposal)

## Checklist

- [x] Added a changelog entry to the `changelog.md` file in either `abc_spec` or `nas_spec`
- [ ] Contacted the Tech Docs team about any corresponding guides that need to be updated for www.checkout.com/docs

**Contributing options**: Look at [our documentation](https://checkout.atlassian.net/wiki/spaces/PD/pages/2169506663/API+ref+publication+process) for options to contribute to the API reference.

### CODE REVIEW

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+type: object
+properties:
+  bin:
+    description: The issuer Bank Identification Number
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:4]:
```suggestion
    description: The issuer's Bank Identification Number (BIN)
```

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    type: string
+    minLength: 6
+    maxLength: 8
+    pattern: '^[0-9]+$'
+    example: '454347'
+  number:
+    description: The PAN (Primary Account Number)
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:11]:
```suggestion
    description: The Primary Account Number (PAN)
```

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    type: string
+    minLength: 12
+    maxLength: 19
+    pattern: '^[0-9]+$'
+    example: '4543474002249996'
+  token_id:
+    description: Reference to CKO token that represents a card's details that can be later used to request a payment
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:18]: So I can suggest a documentation improvement here, can you clarify how `token_id` is different in function to the existing `token` field we have?

The example value provided here seems to mirror those of `token`.

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    type: string
+    pattern: '^(tok)_(\w{26})$'
+    example: 'tok_ubfj2q76miwundwlk72vxt2i7q'
+  instrument_id:
+    description: Reference to instrument
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:23]: Similarly here, is this essentially an alternative to providing an `instrument` object containing an `id` field?

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    type: string
+    pattern: '^(src)_(\w{26})$'
+    example: 'src_e7hqezhxdfeetnyfdv3ayu6yru'
+  format:
+    description: |
+      The output json format. "basic" format means that the response will include only
+      standard metadata. "card_payouts" format means that the response will consist of
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:30]:
```suggestion
```

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    pattern: '^(tok)_(\w{26})$'
+    example: 'tok_ubfj2q76miwundwlk72vxt2i7q'
+  instrument_id:
+    description: Reference to instrument
+    type: string
+    pattern: '^(src)_(\w{26})$'
+    example: 'src_e7hqezhxdfeetnyfdv3ayu6yru'
+  format:
+    description: |
+      The output json format. "basic" format means that the response will include only
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:29]:
```suggestion
      The format to provide the output in. 
      
      A `basic` response will only include standard metadata, while a `card_payouts` formatted response will also include fields specific to card payouts.
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 7]
```diff
@@ -0,0 +1,86 @@
+type: object
+description: Card Metadata
+properties:
+  bin:
+    description: The issuer Bank Identification Number
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:5]:
```suggestion
    description: The issuer's Bank Identification Number (BIN)
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    description: The card category
+    type: string
+    enum: [consumer, commercial, all, other]
+    example: consumer
+  issuer:
+    description: The name of the card issuer
+    type: string
+    example: 'STATE BANK OF MAURITIUS'
+  issuer_country:
+    description: The card issuer country ISO-2 code
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:35]:
```suggestion
    description: The card issuer's country, as an ISO-2 code
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    type: string
+    example: 'MU'
+  issuer_country_name:
+    description: The card issuer country name
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:39]:
```suggestion
    description: The card issuer's country
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    type: string
+    example: 'Mauritius'
+  product_id:
+    description: The issuer/card scheme product identifier
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:43]:
```suggestion
    description: The card issuer or scheme's product identifier
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  product_type:
+    description: The issuer/card scheme product type
+    type: string
+    example: 'F'
+  card_payouts:
+    description: A card payouts block exists in the response if the format in the request is  "card_payouts".
+    type: object
+    properties:
+      domestic_non_money_transfer:
+        description: The attribute shows if the card is eligible for domestic non-money transfer transactions
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:55]:
```suggestion
        description: Describes whether the card is eligible for domestic non-money transfer transactions
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        type: string
+        enum: [not_supported,standard,fast_funds,unknown]
+        example: 'not_supported'
+      cross_border_non_money_transfer:
+        description: The attribute shows if the card is eligible for cross-border non-money transfer transactions.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:60]:
```suggestion
        description: Describes whether the card is eligible for cross-border non-money transfer transactions
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        type: string
+        enum: [not_supported,standard,fast_funds,unknown]
+        example: 'standard'
+      domestic_gambling:
+        description: The attribute shows if the card is eligible for domestic gambling transactions.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:65]:
```suggestion
        description: Describes whether the card is eligible for domestic gambling transactions
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        type: string
+        enum: [not_supported,standard,fast_funds,unknown]
+        example: 'fast_funds'
+      cross_border_gambling:
+        description: The attribute shows if the card is eligible for cross-border gambling transactions.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:70]:
```suggestion
        description: Describes whether the card is eligible for cross-border gambling transactions
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        type: string
+        enum: [not_supported,standard,fast_funds,unknown]
+        example: 'unknown'
+      domestic_money_transfer:
+        description: The attribute shows if the card is eligible for domestic money transfer transactions.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:75]:
```suggestion
        description: Describes whether the card is eligible for domestic money transfer transactions
```

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        type: string
+        enum: [not_supported,standard,fast_funds,unknown]
+        example: 'not_supported'
+      cross_border_money_transfer:
+        description: The attribute shows if the card is eligible for cross-border money transfer transactions
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:80]:
```suggestion
        description: Describes whether the card is eligible for cross-border money transfer transactions
```

File: nas_spec/paths/metadata@card.yaml [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+              token_id: 'tok_ubfj2q76miwundwlk72vxt2i7q'
+              format: basic
+          instrument_example:
+            value:
+              instrument_id: 'src_e7hqezhxdfeetnyfdv3ayu6yru'
+              format: card_payouts
+
+  responses:
+    '200':
+      description: Card metadata for the card number, bin, token or instrument_id provided
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/paths/metadata@card.yaml:38]:
```suggestion
      description: Metadata for the card number, BIN, token, or instrument ID provided
```

File: nas_spec/paths/metadata@card.yaml [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+  tags:
+    - Card metadata
+  security:
+    - OAuth:
+        - vault:card-metadata
+  summary: Get card metadata
+  operationId: requestCardMetadata
+  description: |
+    <span style="font-size: 14px; font-weight: 500; color: rgb(60, 4, 133); background-color: rgb(237, 222, 255); border-radius: 4px; padding: 2px 6px" >Beta</span><p></p>
+    This endpoint provides general information about Cards. Only one parameter must
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/paths/metadata@card.yaml:11]:
```suggestion
    Returns a single metadata record for the card specified by the Primary Account Number (PAN), Bank Identification Number (BIN), token, or instrument ID supplied. 
```

File: nas_spec/paths/metadata@card.yaml [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    be passed as an input parameter in the request. It can be either PAN, BIN, token, or
+    instrument. The endpoint returns one Card Metadata record.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/paths/metadata@card.yaml:13]:
```suggestion
```

File: nas_spec/paths/metadata@card.yaml [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    - Card metadata
+  security:
+    - OAuth:
+        - vault:card-metadata
+  summary: Get card metadata
+  operationId: requestCardMetadata
+  description: |
+    <span style="font-size: 14px; font-weight: 500; color: rgb(60, 4, 133); background-color: rgb(237, 222, 255); border-radius: 4px; padding: 2px 6px" >Beta</span><p></p>
+    This endpoint provides general information about Cards. Only one parameter must
+    be passed as an input parameter in the request. It can be either PAN, BIN, token, or
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/paths/metadata@card.yaml:12]:
```suggestion
```

**Review by sebastian-garcia-cardona-cko — APPROVED** [review]

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+      standard metadata. "card_payouts" format means that the response will consist of
+      standard metadata and Card Payouts specific fields.
```
[the remark this answers is not in the document]
Comment (jaak-niit-cko) [reply nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:31]: Why to delete this line, I guess there is missing something.

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml (hunk 4 above)
Comment (jaak-niit-cko) [reply nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:23]: Yes, its the same.
The instrument_id is currently used in Sessions - https://api-reference.checkout.com/#operation/createSession (in response)

File: nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml (hunk 3 above)
Comment (jaak-niit-cko) [reply nas_spec/components/schemas/Metadata/CardMetadataRequest.yaml:18]: It refers to token that can be requested with this service - https://api-reference.checkout.com/#tag/Tokens
There are 3 options how to provide token as input
1. "token": "tok_4gzeau5o2uqubbk6fufs3m7p54"
2. "token_id": "tok_4gzeau5o2uqubbk6fufs3m7p54"
3. "source": {
    "type": "token",
     "id": "tok_4gzeau5o2uqubbk6fufs3m7p54"
   }
Cko internal API-s currently use the second version and therefore I would prefer this one. Also semantically for me makes sense to have "_id" suffix for fields that refer to other objects.

File: nas_spec/paths/metadata@card.yaml (hunk 19 above)
Comment (jaak-niit-cko) [reply nas_spec/paths/metadata@card.yaml:13]: Sentence will be unfinished if we delete line13.

File: nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+  product_id:
+    description: The issuer/card scheme product identifier
+    type: string
+    example: 'CLASSIC'
+  product_type:
+    description: The issuer/card scheme product type
+    type: string
+    example: 'F'
+  card_payouts:
+    description: A card payouts block exists in the response if the format in the request is  "card_payouts".
```
[the remark this answers is not in the document]
Comment (jaak-niit-cko) [reply nas_spec/components/schemas/Metadata/CardMetadataResponse.yaml:51]: card_payouts object provides merchants visibility of Card Payouts eligibility & fast funds enablement needed before initiating a Card Payout request with CKO.

**Review by sebastian-garcia-cardona-cko — APPROVED** [review]

**Pull request merged**